internal/jobtracker/jobtracker_test.go

```diff
@@ -0,0 +1,189 @@
+//
+// Or we leave it polling.
+func (s *jobTrackerSuite) TestJobTracker_Run_JobIsSetFailed_JobError(c *qt.C) {
+	testCtx := c.Context()
+
+	aFastDyingJob := func(ctx context.Context) error {
+		return errors.New("I died really fast")
+	}
+
+	id, err := s.tracker.Run(testCtx, "test-job-type", aFastDyingJob, time.Second*1000)
+	c.Assert(err, qt.IsNil)
+
+	var status dbmodel.JobStatus
+	var pollerr error
+	for i := 0; i < 10; i++ {
+		status, pollerr = s.db.GetJobStatus(testCtx, id)
+		c.Assert(pollerr, qt.IsNil)
+		if status == dbmodel.StatusFailed {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	j := dbmodel.JobTrackerEntry{}
+	var jobErr string
+	c.Assert(s.db.DB.First(&j).Where("job_id = ?", id).Select("error").Scan(&jobErr).Error, qt.IsNil)
+	c.Assert(jobErr, qt.Equals, "I died really fast")
+}
+
+func (s *jobTrackerSuite) TestJobTracker_Run_JobIsSetFailed_DeadlineExceeded(c *qt.C) {
+	testCtx := c.Context()
+
+	aDeadendJob := func(ctx context.Context) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(10 * time.Second):
+			return nil
+		}
+	}
+
+	id, err := s.tracker.Run(testCtx, "test-job-type", aDeadendJob, time.Millisecond*100)
+	c.Assert(err, qt.IsNil)
+
+	var status dbmodel.JobStatus
+	var pollerr error
+	for i := 0; i < 10; i++ {
+		status, pollerr = s.db.GetJobStatus(testCtx, id)
+		c.Assert(pollerr, qt.IsNil)
+		if status == dbmodel.StatusFailed {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	c.Assert(status, qt.Equals, dbmodel.StatusFailed)
+
+	j := dbmodel.JobTrackerEntry{}
+	var jobErr string
+	c.Assert(s.db.DB.First(&j).Where("job_id = ?", id).Select("error").Scan(&jobErr).Error, qt.IsNil)
+	c.Assert(jobErr, qt.Equals, "context deadline exceeded")
+}
+
+func (s *jobTrackerSuite) TestJobTracker_Run_JobIsSetFailed_CancelledByStop(c *qt.C) {
+	testCtx := c.Context()
+
+	// Create a new tracker, specific to this test.
+	tracker, err := jobtracker.NewJobTracker(s.db, time.Millisecond*50)
+	c.Assert(err, qt.IsNil)
+
+	aStoppedJob := func(ctx context.Context) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(10 * time.Second):
+			return errors.New("timedout")
+		}
+	}
+
+	id, err := tracker.Run(testCtx, "test-job-type", aStoppedJob, time.Second*1000)
+	c.Assert(err, qt.IsNil)
+
+	c.Assert(s.db.StopJob(testCtx, id), qt.IsNil)
+
+	var status dbmodel.JobStatus
+	var pollerr error
+	for i := 0; i < 10; i++ {
+		status, pollerr = s.db.GetJobStatus(testCtx, id)
+		c.Assert(pollerr, qt.IsNil)
+		if status == dbmodel.StatusFailed {
+			break
+		}
+		time.Sleep(10 * time.Second)
+	}
+
+	c.Assert(status, qt.Equals, dbmodel.StatusFailed)
+
+	j := dbmodel.JobTrackerEntry{}
+	var jobErr string
+	c.Assert(s.db.DB.First(&j).Where("job_id = ?", id).Select("error").Scan(&jobErr).Error, qt.IsNil)
+	c.Assert(jobErr, qt.Equals, "context canceled")
+}
+
+func (s *jobTrackerSuite) TestJobTracker_Run_JobIsSetRunning(c *qt.C) {
```

Comment:
Technically this test leaks the go routines managing the job. That is worth addressing.
It also looks like this test could be racy, what if the job completes before you get the job status?

Reply:
It can't complete because the context is alive for the duration of the test?